File: http patch/C#/FundService/FundService/Common/Pipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FundService.Common
{
	public static class Pipeline
	{
		public static Result<TData> RunAll<TData>(TData data, List<Func<TData, Result<TData>>> funcs) {
			var errors = new List<string>();
            var compensators = new List<Action>();
			var currentData = data;
			foreach (var f in funcs) {
				var result = f(currentData);
				if (!result.Success)
					errors.Add(result.Error);
                if (result.ShouldCompensate)
                    compensators.Add(result.Compensate);
				currentData = result.Data;
			}
            if (errors.Any()) {
                compensators.ForEach(c => c());
                new Result<TData>(string.Join("<br/>", errors));
            }
			return new Result<TData>();
		}

		public static Result<TData> RunUntilFirstFault<TData>(TData data, List<Func<TData, Result<TData>>> funcs) {
			foreach (var result in funcs.Select(f => f(data)).Where(result => !result.Success)) {
                result.Compensate();
				return result;
			}
			return new Result<TData>();
		}
	}
}

File: http patch/python/fund/common/pipeline.py

class Result:
    def __init__(self, error=None, data=None, compensate=None):
        self.error = error
        self.data = data
        self.compensate = compensate

    def should_compensate(self):
        return self.compensate is not None

    def success(self):
        return self.error is None


def run_all(data, funcs):
    errors = []
    compensators = []
    current_data = data
    for f in funcs:
        result = f(current_data)
        if not result.success():
            errors.append(result.error)
        current_data = result.data
        if result.should_compensate():
            compensators.append(result.compensate)
    for c in compensators:
        c()
    if len(errors):
        return Result('\n'.join(errors), current_data)
    return Result(data=current_data)


def run_until_first_fault(data, funcs):
    compensators = []
    current_data = data
    for f in funcs:
        result = f(current_data)
        if not result.success():
            return result
        if result.should_compensate():
            compensators.append(result.compensate)
        current_data = result.data
    for c in compensators:
        c()
    return Result(data=current_data)
